Cursor

ali-rostami opened the pull request:
Hi
I know there are things that need to be completed. e.g the description of functions. But I send this pull request to talk about it better.
So let me know what do you think about this @kisvegabor ?
It works for me:
I added these lines to the callback of chart:

```
static void chart_callback(lv_obj_t *obj, lv_event_t event)
{
         if(event == LV_EVENT_PRESSING) {
		lv_indev_t * indev = lv_indev_get_act();
		lv_point_t p;
		lv_indev_get_point(indev, &p);

		lv_area_t chart_coords = {0};
		lv_obj_get_coords(obj, &chart_coords);

		 if(_lv_area_is_point_on(&chart_coords, &p, 0)) {
			 lv_point_t point = {0};
			 point.x = p.x - chart_coords.x1;
			 point.y = p.y - chart_coords.y1;

			 uint16_t index = lv_chart_get_nearest_index_from_coord(chart_trend_spo2, point);
			 point = lv_chart_get_coord_from_index(chart_trend_spo2, series_trend_spo2, index);
			 lv_chart_set_cursor_point(chart_trend_spo2, cursor_trend, point);
			 lv_chart_refresh(obj);
		 }
	}
}
```
And now I can see the cursor when I press the chart.

embeddedt commented on src/lv_widgets/lv_chart.c:
```diff
@@ -702,6 +731,64 @@ lv_chart_axis_t lv_chart_get_series_axis(lv_obj_t * chart, lv_chart_series_t * s
+    return id;
+}
+
+lv_point_t lv_chart_get_coord_from_index(lv_obj_t * chart, lv_chart_series_t * ser, uint16_t id)
```
It might be better to pass a pointer to an `lv_point_t` to this function and fill it inside, rather than returning a struct by value. I think that's how other point-related functions handle it at the moment.

ali-rostami replied:
Yeah, not bad. But as we have 4 inputs I thought It might be the preferred way.
Did you test it @embeddedt ?

ali-rostami replied:
done.

kisvegabor commented on src/lv_widgets/lv_chart.c:
```diff
@@ -702,6 +732,64 @@ lv_chart_axis_t lv_chart_get_series_axis(lv_obj_t * chart, lv_chart_series_t * s
+    } else if(coord.x > w) {
+	    id = ext->point_cnt - 1;
+    } else {
+	    id = coord.x * (ext->point_cnt - 1) / w;
```
Just asking: is it working with line and column type too?

ali-rostami replied:
I forgot the column type, but it seems it works well for line type.

kisvegabor commented on src/lv_widgets/lv_chart.c:
```diff
@@ -702,6 +732,64 @@ lv_chart_axis_t lv_chart_get_series_axis(lv_obj_t * chart, lv_chart_series_t * s
+	lv_coord_t h = lv_area_get_height(&series_area);
+
+	lv_point_t point = {0};
+	point.x = (w * id) / (ext->point_cnt - 1);
```
With column type, it's more complicated as the columns of the same index are drawn tightly next to each other.

ali-rostami replied:
Yeah,  as I said I didn't think about column type.

kisvegabor commented on src/lv_widgets/lv_chart.c:
```diff
@@ -1120,6 +1213,76 @@ static void draw_series_column(lv_obj_t * chart, const lv_area_t * series_area,
+	if(cursor->axis | LV_CHART_CURSOR_DIRECTION_HOR) {
+		p1.x = series_area->x1;
+		p1.y = series_area->y1 + cursor->point.y;
+		p2.x = p1.x + cursor->point.x;
```
So the length of the cursor will be `cursor->point.x`?

ali-rostami replied:
the cursor will be drawn as a solid/dashed line based on its direction from the axes. So when its direction is horizontal yeah it is its length. If it was vertical the point.y became its length. and if we or horizontal and vertical it will draw both of the lines, crossed at the point coordination.

kisvegabor reviewed (commented):
Looks good in general, thank you very much!

A few things:
- Unfortunately, it's more complicated to calculate the position in the case of column type. See my comment.
- I think it's enough to store a position in the cursors descriptor and always draw full widths/height lines.

ali-rostami commented:
> Looks good in general, thank you very much!

Thanks, man.🙏


> * I think it's enough to store a position in the cursors descriptor and always draw full widths/height lines.

Not sure what you mean. would you please explain more. 
Do you mean not defining another style for cursor?
Or you mean it's not a good idea to draw the lines from left and bottom to the point. for this, another idea that comes to my mind is defining secondary direction. I think it's so useful especially for the ones who want to draw the cursor from the secondary Y. I will push new changes regarding this.

ali-rostami commented:
By these new  changes we can have something like this:
![cursor](https://user-images.githubusercontent.com/9710249/93167631-47049a00-f736-11ea-8bac-b7f5bf4860b1.jpeg)

I added the cursor like this:
cursor = lv_chart_add_cursor(chart, LV_COLOR_WHITE,
				LV_CHART_CURSOR_DIRECTION_PRIMARY_HOR | LV_CHART_CURSOR_DIRECTION_PRIMARY_VER | LV_CHART_CURSOR_DIRECTION_SECONDARY_VER);

ali-rostami commented:
This is my code for drawing the column type chart:

```
chart_trend_spo2 = lv_chart_create(cont_middle_menu, NULL);
	lv_obj_set_size(chart_trend_spo2, 698, 240);

	lv_chart_set_type(chart_trend_spo2, LV_CHART_TYPE_COLUMN);
	lv_chart_set_point_count(chart_trend_spo2, 400);
	lv_chart_set_y_range(chart_trend_spo2, LV_CHART_AXIS_PRIMARY_Y, 0,  120);

	series_trend_spo2 = lv_chart_add_series(chart_trend_spo2, LV_COLOR_BLUE);
	for(uint32_t i = 0; i < 100; i++) {
		lv_chart_set_next(chart_trend_spo2, series_trend_spo2, i);
	}

	for(uint32_t i = 100; i > 0; i--) {
		lv_chart_set_next(chart_trend_spo2, series_trend_spo2, i);
	}

	for(uint32_t i = 0; i < 100; i++) {
		lv_chart_set_next(chart_trend_spo2, series_trend_spo2, i);
	}

	for(uint32_t i = 100; i > 0; i--) {
		lv_chart_set_next(chart_trend_spo2, series_trend_spo2, i);
	}
```

Is there any problem with it, cause it doesn't work for me. I use the dev branch right now.

kisvegabor commented on src/lv_widgets/lv_chart.h:
```diff
@@ -122,7 +81,88 @@ enum {
  * @return pointer to the created chart background
  */
-lv_obj_t * lv_chart_create(lv_obj_t * par, const lv_obj_t * copy);
+lv_obj_t * lv_chart_create(lv_obj_t * par, const lv_obj_t * copy)
```
It seems somehow the implementation of the functions are moved to the header file.

ali-rostami replied:
fixed. 
thanks

kisvegabor commented:
Drawing the cursor **to** the point and not in full width/height makes sense if it works with the secondary Y axes too.


The `LV_CHART_CURSOR_DIRECTION_SECONDARY_HOR | LV_CHART_CURSOR_DIRECTION_PRIMARY_VER` names are a little bit misleading to me because we draw a vertical cursor to the horizontal axis. So in `LV_CHART_CURSOR_DIRECTION_PRIMARY_VER` what is `VER`? The cursor or the axis? Maybe left/right/top/bottom would less error-prone. What do you think?

> Is there any problem with it, cause it doesn't work for me. I use the dev branch right now.

It happens because with 400 points the columns will be too narrow (zero pixels wide)

ali-rostami commented:
> Drawing the cursor **to** the point and not in full width/height makes sense if it works with the secondary Y axes too.
Yeah, exactly. 
>  Maybe left/right/top/bottom would less error-prone. What do you think?
I thought about it too, good idea. So you agree to call name it as this:
```LV_CHART_CURSOR_RIGHT
LV_CHART_CURSOR_UP
LV_CHART_CURSOR_LEFT
LV_CHART_CURSOR_DOWN```

> It happens because with 400 points the columns will be too narrow (zero pixels wide)

Oh, I didn't know it. Actually, in ver 6, we could do this. thanks.👍

kisvegabor commented:
I like the new names!

ali-rostami commented:
> I like the new names!

new names are set.

ali-rostami commented:
I tried to test the cursor for the column type and found that I should change the `lv_chart_get_x_from_index` based on the type. But another thing that I saw was If someone wants to draw a chart both as a line and column type -`LV_CHART_TYPE_LINE | LV_CHART_TYPE_COLUMN`- the points of line types will not set in the middle of the columns. which doesn't seem good. 
So, I think we should change the `draw_series_line` function, the same as the `draw_series_column`, somehow we can see the points in the middle of the columns if we or the column type too.
What do you think?

kisvegabor commented:
Thanks for the update!

> But another thing that I saw was If someone wants to draw a chart both as a line and column type -LV_CHART_TYPE_LINE | LV_CHART_TYPE_COLUMN- the points of line types will not set in the middle of the columns. which doesn't seem good.

Normally you can't use both types so I think it's not an issue. 

The other parts look good, but I couldn't test it yet.

ali-rostami commented:
> Normally you can't use both types so I think it's not an issue.

So why not implementing it. It would be great if we could do such a thing as the chart shown above. I think to do this we have to set the type to the series. so a series can be a line type and another one can be a column type.

> The other parts look good, but I couldn't test it yet.
OK, let me know whenever you could 👍

kisvegabor commented:
>  I think to do this we have to set the type to the series. so a series can be a line type and another one can be a column type. what do you think?

Right now `lv_chart_set_type` is set on the chart and not on each series. So it'd be an API break.
I'm not against it at first look but I suggest postponing it until dev-v8 gets more stable.

ali-rostami commented:
> I suggest postponing it until dev-v8 gets more stable.

No problem. 👍

Now the column chart type is supported too and works for me.
If it's OK for you too, tell me to add the descriptions of functions too.
I also have no idea how to solve these:
![image](https://user-images.githubusercontent.com/9710249/93893430-7b5df480-fcfa-11ea-9fc1-309e7e5ad94a.png)
would please guide me.

embeddedt commented:
Here are the errors from CI (you can also see them by clicking Details, opening the failing entry (Run tests), and scrolling through the log):

```
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c: In function ‘lv_chart_get_x_from_index’:
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:772:18: error: ‘return’ with no value, in function returning non-void [-Werror]
  if(ext == NULL) return;
                  ^~~~~~
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:766:12: note: declared here
 lv_coord_t lv_chart_get_x_from_index(lv_obj_t * chart, lv_chart_series_t * ser, uint16_t id)
            ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:773:27: error: ‘return’ with no value, in function returning non-void [-Werror]
  if(id >= ext->point_cnt) return;
                           ^~~~~~
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:766:12: note: declared here
 lv_coord_t lv_chart_get_x_from_index(lv_obj_t * chart, lv_chart_series_t * ser, uint16_t id)
            ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c: In function ‘lv_chart_get_y_from_index’:
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:806:18: error: ‘return’ with no value, in function returning non-void [-Werror]
  if(ext == NULL) return;
                  ^~~~~~
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:800:12: note: declared here
 lv_coord_t lv_chart_get_y_from_index(lv_obj_t * chart, lv_chart_series_t * ser, uint16_t id)
            ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:807:27: error: ‘return’ with no value, in function returning non-void [-Werror]
  if(id >= ext->point_cnt) return;
                           ^~~~~~
/home/runner/work/lvgl/lvgl/tests/../../lvgl/src/lv_widgets/lv_chart.c:800:12: note: declared here
 lv_coord_t lv_chart_get_y_from_index(lv_obj_t * chart, lv_chart_series_t * ser, uint16_t id)
```

ali-rostami commented:
Thanks dear @embeddedt 
so now, what should I return if forexample the id in the `lv_chart_get_x_from_index` function was out of range.
or if the `ext` was null?

embeddedt commented:
I'd just log an error and return 0.

ali-rostami commented:
The cursor positon is based on series area. Not sure if you like it @kisvegabor or not. but if so, I think it's a good idea to make the `get_series_area` function out of static.

kisvegabor commented:
> The cursor positon is based on series area. 

I think series area is fine. For `x = 0` people expect the very left point. 

> I think it's a good idea to make the get_series_area function out of static.

I agree.

ali-rostami commented:
I think it's ready to be reviewed and get merged.

kisvegabor commented:
Thank you, I'll try it out tomorrow.

kisvegabor commented:
Hi,

I've tested this feature and it worked quite well. I've just fixed and changed a few minor things.

If you have no objection against the changes I can merge it.

ali-rostami commented on src/lv_widgets/lv_chart.c:
```diff
@@ -755,30 +756,25 @@ lv_point_t lv_chart_get_cursor_point(lv_obj_t * chart, lv_chart_cursor_t * curso
-	    id = coord.x * (ext->point_cnt - 1) / w;
-    }
+    if(x < 0) return 0;
+    if(x > w) return ext->point_cnt - 1;
```
Just asking:
Why not writing like this:
```
if(x < 0) return 0;
else if(x > w) return ext->point_cnt - 1;
```

kisvegabor replied:
It's kind of subjective but I "believe" in early returns instead of a single return point.
Therefore I return when I can. :slightly_smiling_face:

ali-rostami replied:
OK, so good,
> I "believe" in early returns instead of a single return point.
Hmm, It reminds me Misra C.
By the way, Have you ever thought about passing Misra C?

kisvegabor replied:
I was thinking about it, but now I think that the restrictions are more than the advantages. :slightly_smiling_face:

ali-rostami commented:
Hi
> I've tested this feature and it worked quite well. I've just fixed and changed a few minor things.

Thanks for the fixes you made.🌺🙏

>If you have no objection against the changes I can merge it.

No, It's OK.

kisvegabor commented:
Great!

Only 2 things are missing: updating the CHANGELOG and the dev branch of the docs. Can you help with that too?

ali-rostami commented:
> Great!
> 
> Only 2 things are missing: updating the CHANGELOG and the dev branch of the docs. Can you help with that too?

I would appreciate you doing it yourself. I'm not so familiar with writing these stuffs.

kisvegabor commented:
No problem, I can add it this week.

kisvegabor commented:
I've updated the [docs](https://docs.lvgl.io/dev/en/html/widgets/chart.html#cursor) and merged this improvement.

Thank you very much Ali!
